### PR TITLE
(BKR-871) Remove cygwin installation from package installation

### PR DESCRIPTION
**Pull request by tvpartytonight**
Previous to this commit, beaker would ensure that the cygwin executable
was available prior to package installation and download the executable
if it was not found. Later versions of cygwin fixed a flaw in permissions
for the executable, which broke the way beaker ensured and downloaded
the cygwin executable.

This change removes the notion of checking for cygwin at all. Rather
than conflate the installation of cygwin with package installation,
beaker should separate those concerns. Cygwin should be downloaded and
installed separately from package installation, or handled at the image 
level for the VM used.

Also included in this PR is an optimization for determining the the
architecture of the host; rather than spend time introspecting the
system, a host should just examine its own platform string to determine
the architecture.

[Related JIRA Images ticket](https://tickets.puppetlabs.com/browse/IMAGES-6)

### CODE REVIEW

**Comment by tvpartytonight** [comment]
@puppetlabs/windows this should have some 👀  on please and thank-you.

**Comment by kevpl** [comment]
These changes should be rebased / figured out after #1388 merges, as that work is necessary for Puppet 5 preparations.

**Comment by tvpartytonight** [comment]
@kevpl doesn't look like there were any merge conflicts!

File: lib/beaker/host/windows/pkg.rb [hunk 1]
```diff
@@ -101,15 +89,7 @@ def pe_puppet_agent_promoted_package_info( puppet_collection = nil, opts = {} )
 
   # @api private
   def identify_windows_architecture
-    arch = nil
-    execute("echo '' | wmic os get osarchitecture", :accept_all_exit_codes => true) do |result|
-      arch = if result.exit_code == 0
-        result.stdout =~ /64/ ? '64' : '32'
-      else
-        identify_windows_architecture_from_os_name_for_win2003
```
Comment (Iristyle) [inline lib/beaker/host/windows/pkg.rb]: Should `identify_windows_architecture_from_os_name_for_win2003` be removed as well?

File: lib/beaker/host/windows/pkg.rb [hunk 2]
```diff
@@ -101,15 +89,7 @@ def pe_puppet_agent_promoted_package_info( puppet_collection = nil, opts = {} )
...
-      end
-    end
-    arch
+    platform.arch =~ /64/ ? '64' : '32'
```
Comment (Iristyle) [inline lib/beaker/host/windows/pkg.rb]: How does this work for 3rd parties that don't folllow the same naming conventions?

Seems more robust to just inspect the host.

File: lib/beaker/host/windows/pkg.rb [hunk 3]
```diff
@@ -25,15 +22,6 @@ def install_package(name, cmdline_args = '')
       cygwin = "setup-x86.exe"
     end
 
-    if not check_for_command(cygwin)
```
Comment (Iristyle) [inline lib/beaker/host/windows/pkg.rb]: Is this removed code landing in a new method... or does it already exist elsewhere already?

File: lib/beaker/host/windows/pkg.rb (hunk 2 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/pkg.rb]: 3rd parties can still name images whatever they like; however, platform strings supplied to beaker should still be split on the `variant`-`version`-`arch`.

In this case, if beaker relies on the robust method, you may end up testing the wrong architecture underneath and you would never know unless you were carefully inspecting your testing logs.

File: lib/beaker/host/windows/pkg.rb (hunk 1 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/pkg.rb]: Good catch, this should be removed as well.

File: lib/beaker/host/windows/pkg.rb (hunk 3 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/pkg.rb]: Good point @Iristyle. I think that this points to a guarantee that Windows platforms should have a check for cygwin and install it during initialization. I 'll submit another commit to this PR with those changes and correlating tests.

**Comment by tvpartytonight** [comment]
@jcoconnor I'd be interested if you have any feedback for this PR.

File: lib/beaker/host/windows/pkg.rb (hunk 3 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/pkg.rb]: I've now added functionality to ensure that cygwin is installed using cygcheck during host validation.

**Comment by tvpartytonight** [comment]
I have addressed the feedback and this is ready for further review @kevpl and @Iristyle

File: lib/beaker/host/windows/pkg.rb (hunk 1 above)
Comment (kevpl) [inline lib/beaker/host/windows/pkg.rb]: @tvpartytonight has this been done?

File: lib/beaker/host/windows/pkg.rb [hunk 4]
```diff
@@ -101,23 +89,7 @@ def pe_puppet_agent_promoted_package_info( puppet_collection = nil, opts = {} )
...
-      else
-        identify_windows_architecture_from_os_name_for_win2003
-      end
-    end
-    arch
+    platform.arch =~ /64/ ? '64' : '32'
   end
 
-  # @api private
-  def identify_windows_architecture_from_os_name_for_win2003
```
Comment (tvpartytonight) [inline lib/beaker/host/windows/pkg.rb]: @kevpl that method is deleted here.

File: lib/beaker/host/windows/pkg.rb (hunk 1 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/pkg.rb]: Yes, it has been removed.

File: lib/beaker/host/windows/exec.rb [hunk 5]
```diff
@@ -108,4 +108,13 @@ def selinux_enabled?()
     false
   end
 
+  # Determin if cygwin is actually installed on the SUT.
+  #
+  # @return [Boolean]
+  def cygwin_installed?
+    output = exec(Beaker::Command.new('cygcheck -c cygwin'), :accept_all_exit_codes => true).stdout 
+    return true if output.match(/cygwin/) and output.match(/OK/)
```
Comment (Iristyle) [inline lib/beaker/host/windows/exec.rb]: This should probably be `&&` rather than `and`

File: lib/beaker/host/windows/exec.rb [hunk 6]
```diff
@@ -108,4 +108,13 @@ def selinux_enabled?()
     false
   end
 
+  # Determin if cygwin is actually installed on the SUT.
+  #
+  # @return [Boolean]
+  def cygwin_installed?
+    output = exec(Beaker::Command.new('cygcheck -c cygwin'), :accept_all_exit_codes => true).stdout
```
Comment (Iristyle) [inline lib/beaker/host/windows/exec.rb]: I think `--check-setup` instead of `-c` is more self-documenting

File: lib/beaker/host/windows/pkg.rb (hunk 2 above)
Comment (Iristyle) [inline lib/beaker/host/windows/pkg.rb]: Ok, so that I'm clear, the platform string should have the requested arch in it (which might not match the actual image arch)?  i.e. you may request 32-bit, even though the image is 64-bit.

I asked because the deleted method ` identify_windows_architecture_from_os_name_for_win2003` appeared to be setting arch based on the host capabilities... was its approach flawed then?

File: lib/beaker/host_prebuilt_steps.rb [hunk 7]
```diff
@@ -107,6 +107,7 @@ def validate_host host, opts
         when host['platform'] =~ /cumulus/
           check_and_install_packages_if_needed(host, CUMULUS_PACKAGES)
         when (host['platform'] =~ /windows/ and host.is_cygwin?)
```
Comment (Iristyle) [inline lib/beaker/host_prebuilt_steps.rb]: It seems a little odd to have both `host.is_cygwin?` *and* `host.cygwin_installed?` defined (with a possibility of having `is_cygwin?` be true, but `cygwin_installed?` returning `false`). Do the docs clearly explain the difference?

File: lib/beaker/host/windows/exec.rb (hunk 5 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/exec.rb]: This checks the content of the response because cygcheck still returns a `0` if the package is not found.

Beaker does capture the error level, and you could ostensibly say that by virtue of running `cygcheck` you can confidently say that cygwin itself is installed; I opted for this more verbose method to ensure that the cygwin installation was installed without errors.

Comment (tvpartytonight) [inline lib/beaker/host/windows/exec.rb]: oh, but you are right about using `&&`.

File: lib/beaker/host/windows/exec.rb (hunk 6 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/exec.rb]: Updated to use `--check-setup`.

File: lib/beaker/host_prebuilt_steps.rb (hunk 7 above)
Comment (tvpartytonight) [inline lib/beaker/host_prebuilt_steps.rb]: Updated for a more verbose comment for `cygwin_installed?` and to point out that `is_cygwin?` is just a type check for `Windows::Host`.

File: lib/beaker/host/windows/pkg.rb (hunk 2 above)
Comment (tvpartytonight) [inline lib/beaker/host/windows/pkg.rb]: I think that approach was flawed for the reason stated above; a platform string must have an arch defined, and there seems to be no good reason to introspect the system for it when it is already an attribute of the host construct itself.

**Comment by tvpartytonight** [comment]
@puppetlabs-jenkins retest this please

**Comment by tvpartytonight** [comment]
@Iristyle I think I have addressed your concerns; can you take another look?

File: lib/beaker/host/windows/exec.rb [hunk 8]
```diff
@@ -108,4 +108,14 @@ def selinux_enabled?()
     false
   end
 
+  # Determine if cygwin is actually installed on the SUT. Differs from
+  # is_cygwin?, which is just a type check for a Windows::Host.
+  #
+  # @return [Boolean]
+  def cygwin_installed?
+    output = exec(Beaker::Command.new('cygcheck --check-setup cygwin'), :accept_all_exit_codes => true).stdout 
+    return true if output.match(/cygwin/) and output.match(/OK/)
```
Comment (Iristyle) [inline lib/beaker/host/windows/exec.rb]: Based on your comment I think you meant to change `and` to `&&`?

Re: exit codes.  You checked multiple Cygwin versions and they all respond the same way? When the check fails what if it responds with the string `NOT OK`?  I just figured the exit code check would insulate you from such differences.  If the binary is not found it should yield `127` (command not found), else the value from `cygcheck`

Comment (Iristyle) [inline lib/beaker/host/windows/exec.rb]: To make my suggestion more concrete:

```ruby
def cygwin_installed?
  response = exec(Beaker::Command.new('cygcheck --check-setup cygwin'), :accept_all_exit_codes => true)

  # command not found (127) or cygcheck failure code indicate not installed
  response.exit_code == 0
end
```

Comment (Iristyle) [inline lib/beaker/host/windows/exec.rb]: So it looks like `cygcheck` always returns `0`, regardless of whether or not it finds a package by the given name.  Oh well....

**Comment by Iristyle** [comment]
:shipit:
